### PR TITLE
`ServerDriveQueryDirectoryRequest` and `SharedDirectoryListRequest/Response`

### PR DESCRIPTION
**Pull request by ibeckermayer**
Corresponds with https://github.com/Devolutions/IronRDP/pull/288

### CODE REVIEW

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr/filesystem.rs [hunk 1]
```diff
@@ -268,6 +273,85 @@ impl FilesystemBackend {
...
+                            dir.contents = tdp_resp.fso_list;
+                            // And send back the "." directory over RDP
+                            return cli.send_next_drive_query_dir_response(&rdp_req);
+                        }
+
+                        cli.send_drive_query_dir_response(
+                            rdp_req.device_io_request,
+                            NtStatus::UNSUCCESSFUL,
+                            None,
+                        )
```
Comment (probakowski) [inline lib/srv/desktop/rdp/rdpclient/src/rdpdr/filesystem.rs:341]: Can we extract this to separate method? It's really hard to read when it's so deeply nested

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr/filesystem.rs [hunk 2]
```diff
@@ -471,6 +569,30 @@ impl FilesystemBackend {
...
 
+    /// Called from the Go code when a [`tdp::SharedDirectoryListResponse`] is received from the browser.
+    ///
+    /// Calls the [`SharedDirectoryListResponseHandler`] associated with the completion id of the
+    /// [`tdp::SharedDirectoryListResponse`].
+    pub fn handle_tdp_sd_list_response(
+        &mut self,
+        tdp_resp: tdp::SharedDirectoryListResponse,
+    ) -> PduResult<()> {
+        if let Some(handler) = self
```
Comment (probakowski) [inline lib/srv/desktop/rdp/rdpclient/src/rdpdr/filesystem.rs:580]: `match` here might be nicer

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr/filesystem.rs (hunk 1 above)
Comment (zmb3) [reply lib/srv/desktop/rdp/rdpclient/src/rdpdr/filesystem.rs:341]: +1, and consider handling the error case (file not found in cache) up front.

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr/filesystem.rs (hunk 2 above)
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/src/rdpdr/filesystem.rs:580]:
```rust
    /// Called from the Go code when a [`tdp::SharedDirectoryListResponse`] is received from the browser.
    ///
    /// Calls the [`SharedDirectoryListResponseHandler`] associated with the completion id of the
    /// [`tdp::SharedDirectoryListResponse`].
    pub fn handle_tdp_sd_list_response(
        &mut self,
        tdp_resp: tdp::SharedDirectoryListResponse,
    ) -> PduResult<()> {
        match self
            .pending_sd_list_resp_handlers
            .remove(&tdp_resp.completion_id)
        {
            Some(handler) => handler.call(self, tdp_resp),
            None => Err(custom_err!(
                "FilesystemBackend::handle_tdp_sd_list_response",
                FilesystemBackendError(format!(
                    "received invalid completion id: {}",
                    tdp_resp.completion_id
                ))
            )),
        }
    }
```

Agreed this is a bit easier to read. This pattern is used repeatedly throughout the stack, so I'll add a PR at the tip that refactors all of them.

File: lib/srv/desktop/rdp/rdpclient/src/rdpdr/filesystem.rs (hunk 1 above)
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/src/rdpdr/filesystem.rs]: b4518cd604b3c2b969ee16488e1d72118c4c0203

File: lib/srv/desktop/rdp/rdpclient/src/client.rs [hunk 3]
```diff
@@ -820,6 +841,21 @@ impl ClientHandle {
...
     }
 
+    pub fn handle_tdp_sd_list_response(
+        &self,
+        res: tdp::SharedDirectoryListResponse,
+    ) -> ClientResult<()> {
+        self.blocking_send(ClientFunction::HandleTdpSdListResponse(res))
+    }
+
+    pub async fn handle_tdp_sd_list_response_async(
```
[the remark this answers is not in the document]
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/src/client.rs:851]: It's never used, I was just being comprehensive in terms of building out the `ClientHandle`'s API.

**Review by probakowski — APPROVED** [review]

**Pull request merged**